### DISCUSSION TITLE
Where is the payload of Rest API endpoints ?

### DISCUSSION THREAD

**Discussion by jmurrayrs** [keycloak/keycloak#29128] [Q&A] [answered]
I'm working on a C# Nuget package project, but I'm having trouble finding documentation on the expected payloads for the endpoints listed in the '.well-known/openid-configuration' file. Where can I find this information?

Thanks!

**Comment by codespearhead** [comment]
Those endpoints adhere to the [OAuth2](https://datatracker.ietf.org/doc/html/rfc6749) and [OpenID Connect](https://openid.net/developers/specs/) standards, so it doesn't make sense for Keycloak or any other standards-compliant server or client to document them.

I can't tell from your question what you're trying to develop, but if that is an OIDC-compliant client library (see [Certified Relying Party Libraries](https://openid.net/developers/certified-openid-connect-implementations/)), then I suggest using the [official Microsoft one](https://www.nuget.org/packages/OpenIddict.AspNetCore).

**Reply by jmurrayrs** [reply]
Hello @codespearhead. Thanks for your help. However, that seems to be more of a personal perspective. We have a lot of services that implements a lot of document standarts and we have access to documentation. It's necessary documentation to keycloak too. But, I think your answer solve my question. It seems that the documentation payload doesn't exists, don't you ?

**Reply by codespearhead** [reply]
Before those standards came to be, every auth solution had its own way of doing things, in which case documenting them would make sense, especially because each one of them would have to provide (or offer thorough documentation about their implementation for third-parties to create) their own client to work with that specific provider.

Now any OIDC client will be able to interact with any OIDC-compliant server, without them knowing which server or client they are interacting with so long as both of them are [certified](https://openid.net/certification/), so the whole complexity revolving around implementation details is abstracted away and documented in a centralized manner in those specs, allowing you, the end user, to just use them instead of having to worry about their internals.

That's why I wondered about what you're trying to accomplish, because if we have the bigger picture, it's very likely there's documention about the how to accomplish it in a declarative way.

However, if you want to learn about the internals to, let's say, contribute to the spec itself or to the codebase of certified implementations, then there are some [alternatives tutorials in the wild that answer your question in a concise manner](https://connect2id.com/products/server/docs/api), but there's no guarantee they're up to date with the specs.

**Reply by jmurrayrs** [reply]
@codespearhead , I'm creating a Nuget Package that serves as an authentication proxy, so the developer won't need direct access to Keycloak. The best (and cheapest) way to do this is to provide a library where the developer can configure their realm, client, and client_secret to access Keycloak. This package do communication with Keycloak using RestAPI endpoints ...

**Reply by codespearhead** [reply]
In that case, does the [Keycloak Admin REST API](https://www.keycloak.org/docs-api/latest/rest-api/index.html) fit the bill?

There's this ongoing initiative from the Keycloak community about documenting relevant spec endpoints used by that REST API in a more succinct way, akin to the [OWASP Cheat Sheet](https://cheatsheetseries.owasp.org/index.html).

It includes OpenAPI definitions in the section of the same name, which you can visualize for convenience in the [OpenAPI UI](https://editor.swagger.io/).

**Reply by jmurrayrs** [reply]
I thought this [documentation](https://www.keycloak.org/docs-api/latest/rest-api/index.html) a little confused. For example I didn't find the **/realms/{realm}/protocol/openid-connect/token** payload. I spent a bunch of time to get this information, being that in theory the documentation would have it, or I'm looking in the wrong place.

**Comment by jmurrayrs** [comment, marked as answer]
Unfortunately, I needed to do this. I pasted the routes into ChatGPT, and it gave me the payloads. So, I created a public repository with JSON files, which can be downloaded using Postman or Insomnia.

"Intelligence sometimes plays tricks on our ego, and it responds with arrogance. Be empathetic. It's a sign of humanity!"

Julien Nascimento


https://github.com/jmurrayrs/keycloak
